<div dir="rtl" style="text-align:right">
<h1 id="מחברת-מלאה-להרצאה-2">מחברת מלאה להרצאה — 2</h1>
<p dir="rtl" style="text-align:right">נקרא את ההסבר, נחזה תוצאה, נריץ את התא הבא ונפרש את הפלט לפני שנמשיך. נשתמש ב־<strong>Restart Kernel and Run All</strong> כדי לוודא שהעבודה אינה תלויה בהרצה ישנה. ההסברים, הגזירות המתמטיות וההנחות הפיסיקליות מהפרק נשמרים סביב התאים.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="הכנה-מסופקת-למחברת">הכנה מסופקת למחברת</h2>
<p dir="rtl" style="text-align:right">נבחר ליבה <strong>Julia 1.12</strong> ונפתח את המחברת מתוך תיקיית הפרק שלה. ההכנה מאתרת את סביבת הקורס המתועדת ויוצרת תיקייה נפרדת לאיורים; היא אינה מתקינה חבילות. אלה פקודות תשתית מסופקות, ולא חומר חדש להערכה.</p>
</div>

In [1]:
import Pkg
course_root = abspath("../..")
if !isfile(joinpath(course_root, "Project.toml"))
    error("Start this notebook from its week folder, e.g. week-1/notebooks")
end
Pkg.activate(course_root; io=devnull)
output_dir = get(ENV, "COURSE_NOTEBOOK_OUTPUT", joinpath(pwd(), "generated", "lecture-he"))
mkpath(output_dir)
println("Course environment active; figure output directory ready.")

Course environment active; figure output directory ready.


<div dir="rtl" style="text-align:right">
<h1 id="ערכים-החלטות-וגוף-על-מישור-משופע">ערכים, החלטות וגוף על מישור משופע</h1>
<h2 id="הרצאה-ממשתנים-לאלגוריתם">הרצאה: ממשתנים לאלגוריתם</h2>
<p dir="rtl" style="text-align:right">בשבוע 1 נתנו שמות למדידות וחישבנו מהן ערכים חדשים. לפני שנוסיף החלטות ולולאות, נאט וננסח במפורש את הרעיון הבסיסי הזה: מהו משתנה, מה עושה פעולת ההצבה וכיצד ערכים משתנים כאשר <code dir="ltr">Julia</code> מבצעת שורה אחר שורה.</p>
<h3 id="משתנה-נותן-שם-לערך">משתנה נותן שם לערך</h3>
<p dir="rtl" style="text-align:right">משתנה הוא שם שמפנה לערך. שם מועיל מאפשר לנו לדבר על משמעותו של מספר במקום לכתוב שוב ושוב את המספר עצמו:</p>
<p dir="rtl" style="text-align:right">להצבה יש צורה כללית <code dir="ltr">name = expression</code>. תחילה <code dir="ltr">Julia</code> מחשבת את הביטוי שמימין, ולאחר מכן השם שמשמאל מפנה לתוצאה. הדוגמה הקטנה ביותר היא <code dir="ltr">x = 2</code>. חישוב מדעי עשיר יותר משתמש בכמה שמות בעלי משמעות:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [2]:
mass = 2.0        # kg
speed = 3.0       # m/s
momentum = mass * speed   # kg m/s

println(momentum)   # 6.0

6.0


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">6.0</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">סימן השוויון היחיד הוא <em>אופרטור ההצבה</em>. תחילה <code dir="ltr">Julia</code> מחשבת את הביטוי שמימין ל-<code dir="ltr">=</code>, ולאחר מכן השם שמשמאל מפנה לערך שהתקבל. לכן בשורה השלישית התוכנית כופלת תחילה את <code dir="ltr">mass</code> ב-<code dir="ltr">speed</code>, ואז מציבה את התוצאה ב-<code dir="ltr">momentum</code>.</p>
<h4 id="תנע-של-מסה-נעה">תנע של מסה נעה</h4>
<p dir="rtl" style="text-align:right">התנע של גוף מודד כמה תנועה הוא נושא: גוף כבד יותר וגוף מהיר יותר נושאים כל אחד יותר תנועה. עבור מסה נקודתית הנעה הרבה מתחת למהירות האור, <span dir="ltr" class="math inline">\(\mathbf{p}=m\mathbf{v}\)</span>, ובגודל <span dir="ltr" class="math inline">\(p=mv\)</span> לאורך כיוון התנועה. המסה <span dir="ltr" class="math inline">\(m\)</span> בקילוגרמים, המהירות <span dir="ltr" class="math inline">\(v\)</span> במטר לשנייה, והתנע בקילוגרם-מטר לשנייה (<span dir="ltr" class="math inline">\(\mathrm{kg\,m/s}\)</span>).</p>
<p dir="rtl" style="text-align:right">אפשר לבדוק משתנה על ידי כתיבת שמו בתא של מחברת, או להדפיס אותו כחלק מהודעה:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אלה ביטויים אינטראקטיביים נפרדים. נציב כל אחד בתא משלו כדי לראות את כל התוצאות ולא רק את האחרונה.</p>
</div>

In [3]:
momentum

6.0

In [4]:
println("momentum = $(momentum) kg m/s")

momentum = 6.0 kg m/s


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">במחברת השורה הראשונה מציגה <code dir="ltr">6.0</code>; הדוח המודפס הוא:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">momentum = 6.0 kg m/s</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="התוכנית-מתבצעת-לפי-הסדר">התוכנית מתבצעת לפי הסדר</h3>
<p dir="rtl" style="text-align:right">הצבה מבצעת חישוב ברגע שבו מגיעים אליה. היא אינה יוצרת קשר אלגברי קבוע בין משתנים:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [5]:
x = 4
y = x + 1       # y is now 5
x = 10          # x changes, but y is still 5

10

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">כאשר <code dir="ltr">Julia</code> מבצעת את <code dir="ltr">y = x + 1</code>, היא משתמשת בערך הנוכחי של <code dir="ltr">x</code>. שינוי מאוחר יותר של <code dir="ltr">x</code> אינו מחשב מחדש את <code dir="ltr">y</code>; לשם כך צריך לבצע שוב את <code dir="ltr">y = x + 1</code>. לסדר השורות חשיבות מיוחדת במחברות, שבהן אפשר להריץ תא יותר מפעם אחת או בסדר שונה.</p>
<h3 id="אפשר-להציב-במשתנה-ערך-חדש">אפשר להציב במשתנה ערך חדש</h3>
<p dir="rtl" style="text-align:right">אותו שם יכול להפנות לערך חדש בהמשך התוכנית:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [6]:
temperature = 20.0   # C
temperature = temperature + 0.5

println(temperature)     # 20.5

20.5


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">20.5</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">השורה השנייה אינה משוואה אלגברית. <code dir="ltr">Julia</code> קוראת את הערך הישן בצד ימין, מוסיפה <code dir="ltr">0.5</code>, ומציבה את התוצאה החדשה בחזרה באותו שם. לולאות ישתמשו בדפוס זה כדי לקדם זמן, להגדיל מונה או לעדכן מקום בסימולציה.</p>
<h4 id="שמות-צריכים-לבטא-משמעות">שמות צריכים לבטא משמעות</h4>
<p dir="rtl" style="text-align:right"><strong>שמות</strong>: שמות משתנים מבחינים בין אותיות גדולות לקטנות, אינם יכולים להכיל רווחים, ויש להשתמש בהם בעקביות. אפשר להשתמש ב-<code dir="ltr">unicode</code>, ובמיוחד באותיות יווניות (למשל להקליד <code dir="ltr">\phi&lt;Tab&gt;</code> כדי לקבל <span dir="ltr" class="math inline">\(\phi\)</span>). נעדיף שמות בעלי משמעות על פני שמות לא מוסברים כגון <code dir="ltr">a</code>, <code dir="ltr">b</code> ו-<code dir="ltr">c</code>. <strong>יחידות</strong>: <code dir="ltr">Julia</code> שומרת בדרך כלל מספרים רגילים בלי היחידות שלהם. בפיסיקה, לעומת זאת, כדאי לעקוב אחריהן. אפשר להוסיף סיומות יחידה לשמות, למשל <code dir="ltr">m_kg</code>, או לחלופין, כפי שאנו עושים כאן, לתעד את היחידה בהערה קצרה בסוף השורה במקום שבו הערך מוגדר, למשל <code dir="ltr">time = 0.35 # s</code>, <code dir="ltr">height = 1.20 # m</code> ו-<code dir="ltr">g = 9.81 # m/s^2</code>; זה אינו חובה אך שימושי מאוד כדי להימנע משגיאות יחידות. במטלת הבונוס 2 (כלומר, שאינה חלק מחומר הקורס הרשמי) נראה שאפשר להשתמש בחבילה <code dir="ltr">Unitful</code> שמוסיפה יחידות למשתנים ומאפשרת לעקוב אחריהן. יש לציין יחידות גם במחרוזות התיעוד או בתיעוד הממשק של פונקציות, ולהבהיר אותן באתרי קריאה שבהם הן עלולות להיות עמומות.</p>
<h3 id="לערכים-יש-טיפוסים">לערכים יש טיפוסים</h3>
<p dir="rtl" style="text-align:right">משתנים יכולים לתת שמות לסוגים שונים של ערכים. בשבוע 1 כבר השתמשנו במספר שלם לספירה, במספרים עשרוניים למדידות ובתוויות טקסט. השבוע נוסיף תוצאות אמת ושקר של השוואות ונציג ארבע הבחנות במפורש, משום שתוכנית שקוראת קלט, בוחרת ענף ומחליטה מתי לעצור חייבת לדעת באיזה סוג ערך היא מטפלת.</p>
<p dir="rtl" style="text-align:right"><em>טיפוס</em> הוא קטגוריה של ערכים שקובעת כיצד <code dir="ltr">Julia</code> שומרת אותם ואילו פעולות מתאימות להם. למשל, ל-<code dir="ltr">5</code>, ל-<code dir="ltr">0.35</code>, ל-<code dir="ltr">"height"</code> ול-<code dir="ltr">true</code> יש צורות שונות, והם מייצגים מספר שלם, מספר נקודה צפה, טקסט וערך אמת או שקר.</p>
<h3 id="ארבעה-טיפוסי-ערכים-פשוטים">ארבעה טיפוסי ערכים פשוטים</h3>
<h4 id="ארבעת-הטיפוסים-הדרושים-כעת">ארבעת הטיפוסים הדרושים כעת</h4>
<table>
<thead>
<tr>
<th style="text-align: left;">טיפוס ב-<code dir="ltr">Julia</code></th>
<th style="text-align: left;">דוגמה</th>
<th style="text-align: left;">תפקיד מדעי נפוץ</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code dir="ltr">Int</code></td>
<td style="text-align: left;"><code dir="ltr">n = 5</code></td>
<td style="text-align: left;">ספירה ואינדקס של לולאה</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">Float64</code></td>
<td style="text-align: left;"><code dir="ltr">t = 0.35 # s</code></td>
<td style="text-align: left;">מדידות וגדלים מחושבים</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">String</code></td>
<td style="text-align: left;"><code dir="ltr">label = "height [m]"</code></td>
<td style="text-align: left;">קלט טקסטואלי, תוויות והודעות</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">Bool</code></td>
<td style="text-align: left;"><code dir="ltr">ready = true</code></td>
<td style="text-align: left;">תשובות אמת/שקר לתנאים</td>
</tr>
</tbody>
</table>
<p dir="rtl" style="text-align:right">הפונקציה <code dir="ltr">typeof(value)</code> מאפשרת לבדוק את טיפוס הערך. משתמשים בה לאבחון, ולא כתחליף להבנת משמעות הערך.</p>
<p dir="rtl" style="text-align:right">במחשבי הקורס הרגילים בעלי 64 סיביות, <code dir="ltr">typeof(5)</code> מדפיסה <code dir="ltr">Int64</code>. השם <code dir="ltr">Int</code> ב-<code dir="ltr">Julia</code> מציין את גודל המספר השלם הרגיל של המחשב, ובמחשבים אלה הוא שם חלופי ל-<code dir="ltr">Int64</code>.</p>
<p dir="rtl" style="text-align:right">הטיפוסים קשורים ישירות לשבוע 1:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [7]:
current = [0.10, 0.20, 0.30, 0.40, 0.50]   # A
n = length(current)                 # Int
I_bar = sum(current) / n            # Float64, A
axis_label = "current I [A]"    # String
analysis_ready = true           # Bool

println(typeof(n))
println(typeof(I_bar))
println(typeof(axis_label))
println(typeof(analysis_ready))

Int64


Float64
String
Bool


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">מספר שלם ומספר נקודה צפה יכולים להשתתף באותו חישוב. <code dir="ltr">Julia</code> ממירה את המספר השלם לפי הצורך:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אלה ביטויים אינטראקטיביים נפרדים. נציב כל אחד בתא משלו כדי לראות את כל התוצאות ולא רק את האחרונה.</p>
</div>

In [8]:
2 * 0.35       # 0.7, a Float64

0.7

In [9]:
5 / 2          # 2.5, a Float64

2.5

In [10]:
Float64(5)     # 5.0

5.0

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">כללים מעשיים אלה מספיקים כעת. אין צורך בהצהרות טיפוס, בטיפוסים מופשטים או בפרטי הייצוג של מספרי נקודה צפה.</p>
<h4 id="טיפוס-אינו-יחידה-פיסיקלית">טיפוס אינו יחידה פיסיקלית</h4>
<p dir="rtl" style="text-align:right">ב-<code dir="ltr">t = 0.35 # s</code>, הטיפוס של <code dir="ltr">Julia</code> הוא <code dir="ltr">Float64</code>; היחידה הפיסיקלית היא שניות. היחידה נמצאת בהערה ובתיעוד הממשק הרלוונטי, ולכן ערך אחר מאותו טיפוס <code dir="ltr">Float64</code> יכול לייצג מטרים או קילוגרמים. בדיקת טיפוסים וניתוח ממדים עונים על שאלות שונות, ואנו זקוקים לשניהם.</p>
<h3 id="השוואות-יוצרות-ערכי-אמת-ושקר">השוואות יוצרות ערכי אמת ושקר</h3>
<p dir="rtl" style="text-align:right">תוכנית זקוקה לשאלות שהתשובה עליהן כן או לא כדי לבדוק נתונים ולבחור מה לעשות בהמשך. להשוואה יש צורה <code dir="ltr">left_value operator right_value</code>, והיא מחזירה <code dir="ltr">Bool</code>: הערך <code dir="ltr">true</code> או <code dir="ltr">false</code>. נתחיל בשאלות ישירות שתשובותיהן ברורות:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אלה ביטויים אינטראקטיביים נפרדים. נציב כל אחד בתא משלו כדי לראות את כל התוצאות ולא רק את האחרונה.</p>
</div>

In [11]:
2 < 3          # true

true

In [12]:
2 == 3         # false

false

In [13]:
5 != 4         # true

true

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אופרטורי ההשוואה העיקריים הם <code dir="ltr">&lt;</code>, <code dir="ltr">&lt;=</code>, <code dir="ltr">&gt;</code>, <code dir="ltr">&gt;=</code>, <code dir="ltr">==</code> (שווה) ו-<code dir="ltr">!=</code> (אינו שווה). סימן שוויון יחיד <code dir="ltr">=</code> מציב ערך; שני סימני שוויון <code dir="ltr">==</code> שואלים שאלה.</p>
<p dir="rtl" style="text-align:right">אפשר לשלב ערכי אמת ושקר כאשר השוואה אחת אינה מספיקה. האופרטורים <code dir="ltr">&amp;&amp;</code>, <code dir="ltr">||</code> ו-<code dir="ltr">!</code> פירושם וגם, או, ולא:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אלה ביטויים אינטראקטיביים נפרדים. נציב כל אחד בתא משלו כדי לראות את כל התוצאות ולא רק את האחרונה.</p>
</div>

In [14]:
true && false      # false: both must be true

false

In [15]:
true || false      # true: at least one is true

true

In [16]:
!true              # false

false

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אותן צורות יכולות לבטא כלל פיסיקלי עשיר יותר לבדיקת תקינות:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [17]:
t = 0.35   # s
z0 = 1.20   # m
nonnegative_time = t >= 0
valid_height = z0 >= 0
valid_input = nonnegative_time && valid_height

true

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="החלטה-ראשונה-בעזרת-if">החלטה ראשונה בעזרת if</h3>
<p dir="rtl" style="text-align:right">חישוב ערך אמת או שקר מועיל רק אם התוכנית יכולה לפעול לפיו. הוראת <code dir="ltr">if</code> היא תנאי: היא מריצה בלוק אחד כאשר התנאי <code dir="ltr">true</code>. בלוק <code dir="ltr">else</code> אופציונלי רץ כאשר התנאי <code dir="ltr">false</code>. הצורה הכללית היא:</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="תבנית-תחביר">תבנית תחביר</h3>
<p dir="rtl" style="text-align:right">זו הצורה הכללית; <code dir="ltr">condition</code> הוא מציין מקום. נקרא אותה כאן ואז נריץ את הדוגמה המוחשית שאחריה.</p>
</div>

In [ ]:
if condition
    # commands when the condition is true
else
    # commands when the condition is false
end

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">כל תנאי מסתיים ב-<code dir="ltr">end</code>. אם אין צורך לעשות דבר במקרה השקרי, משמיטים את בלוק ה-<code dir="ltr">else</code>. הנה הדוגמה הדו-ענפית הקטנה ביותר:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [18]:
temperature = 25.0   # C
if temperature > 30
    println("hot")
else
    println("not hot")
end

not hot


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">not hot</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">לעתים המשך החישוב ייצור תוצאה חסרת משמעות. <code dir="ltr">Julia</code> מספקת עצירה מכוונת בצורה זו:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="שגיאה-מכוונת-בלי-לעצור-את-המחברת">שגיאה מכוונת, בלי לעצור את המחברת</h3>
<p dir="rtl" style="text-align:right">כאן הפרק מדגים עצירה מכוונת. עטיפת <code dir="ltr">try</code>/<code dir="ltr">catch</code> המסופקת מדפיסה את השגיאה הצפויה ומאפשרת להמשיך בהרצת כל התאים; שגיאה אחרת אינה מוסתרת. נתמקד בהוראה שנכשלה ולא בלימוד העטיפה בשלב זה.</p>
</div>

In [19]:
try
    throw(ArgumentError("message"))
catch err
    err isa ArgumentError || rethrow()
    showerror(stdout, err)
    println()
end

ArgumentError: message


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">היא עוצרת את הביצוע ומדווחת מדוע. נשתמש בה לאחר המרה שנכשלה ולאחר קלט שאינו חוקי מבחינה פיסיקלית.</p>
<h2 id="קינמטיקה-בחלקים-כהחלטה">קינמטיקה בחלקים כהחלטה</h2>
<h4 id="תאוצה-מסומנת-לעומת-גודל-הכבידה-החיובי">תאוצה מסומנת לעומת גודל הכבידה החיובי</h4>
<p dir="rtl" style="text-align:right">גובה חיובי כלפי מעלה, ולכן תאוצת הכובד היא <span dir="ltr" class="math display">\[g=-9.81\ \mathrm{m/s^2}.\]</span> עבור גובה התחלתי <span dir="ltr" class="math inline">\(z_0\)</span> ומהירות התחלתית <span dir="ltr" class="math inline">\(v_0\)</span>, <span dir="ltr" class="math display">\[z_{\mathrm{free}}(t)=z_0+v_0t+\frac{1}{2}gt^2.\]</span> לשלושת האיברים יחידות של מטרים. בדיקת ממדים זו אינה תלויה בכך שערכי הקוד הם מטיפוס <code dir="ltr">Float64</code>.</p>
<p dir="rtl" style="text-align:right">המודל ללא קפיצה אינו מאפשר לכדור להמשיך מתחת לקרקע: <span dir="ltr" class="math display">\[z(t)=\begin{cases}
z_{\mathrm{free}}(t),&amp;z_{\mathrm{free}}(t)&gt;0,\\
0,&amp;z_{\mathrm{free}}(t)\leq0.
\end{cases}\]</span></p>
<p dir="rtl" style="text-align:right">ההגדרה בחלקים מתורגמת ישירות לתנאי:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [20]:
t = 0.35   # s
z0 = 1.20   # m
v0 = 0.0    # m/s
g = -9.81   # m/s^2

if t < 0 || z0 < 0
    throw(ArgumentError("time and initial height must be non-negative"))
end

z_free = z0 + v0 * t + 0.5 * g * t^2   # m

if z_free > 0
    z = z_free
    state = "in air"
else
    z = 0.0
    state = "on the ground"
end

println("at t = $(t) s: z = $(round(z, digits=3)) m ($(state))")

at t = 0.35 s: z = 0.599 m (in air)


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">at t = 0.35 s: z = 0.599 m (in air)</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">ב-<span dir="ltr" class="math inline">\(t=0.35\ \mathrm{s}\)</span> התוכנית מדווחת בקירוב <span dir="ltr" class="math inline">\(z=0.599\ \mathrm{m}\)</span>. התוצאה עוברת שלוש בדיקות: הזמן בתחום המודל, הגובה נמוך מן הגובה ההתחלתי, ולכל איבר במשוואת המקום יחידות של מטרים.</p>
<h3 id="כמה-ענפים">כמה ענפים</h3>
<p dir="rtl" style="text-align:right">משתמשים ב-<code dir="ltr">elseif</code> כאשר יש יותר משני מקרים זרים זה לזה שיש לבדוק לפי סדר. הצורה הכללית היא <code dir="ltr">if</code>, אחריו מספר כלשהו של ענפי <code dir="ltr">elseif</code>, ענף <code dir="ltr">else</code> אופציונלי, ולבסוף <code dir="ltr">end</code> אחד. רק הענף האמיתי הראשון רץ. דוגמה קטנה לסיווג סימן מציגה את המבנה:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [21]:
value = -2
if value < 0
    description = "negative"
elseif value == 0
    description = "zero"
else
    description = "positive"
end

"negative"

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">אותו מבנה נותן סיווג שימושי יותר של גובה:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [22]:
if z == 0
    state = "on the ground"
elseif z < 0.20
    state = "close to the ground"
else
    state = "in air"
end

"in air"

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">לעתים דרושות החלטות מקוננות, אך שרשרת שטוחה של <code dir="ltr">if</code>/<code dir="ltr">elseif</code>/<code dir="ltr">else</code> קלה יותר לבדיקה כאשר היא מבטאת אותו היגיון.</p>
<h4 id="שלוש-קטגוריות-של-שגיאות">שלוש קטגוריות של שגיאות</h4>
<ul>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>שגיאת תחביר:</strong> <code dir="ltr">Julia</code> אינה יכולה לפענח את התוכנית, למשל בגלל <code dir="ltr">end</code> או סוגר חסרים.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>שגיאת זמן ריצה:</strong> התוכנית מתחילה ואז מגיעה לפעולה לא חוקית, למשל העלאת <code dir="ltr">String</code> בחזקה.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>שגיאה לוגית:</strong> התוכנית רצה אך מממשת אלגוריתם או פיסיקה שגויים, למשל <span dir="ltr" class="math inline">\(g\)</span> חיובי כאשר גובה חיובי כלפי מעלה, או זמן שאינו מתקדם בלולאת <code dir="ltr">while</code>.</p></li>
</ul>
<p dir="rtl" style="text-align:right">קוראים את השגיאה המדויקת ואת מספר השורה, מבודדים את הבלוק הקטן ביותר שנכשל, בודקים ערכים וטיפוסים, בודקים ממדים וסימנים, ומריצים מקרים שתשובתם ידועה.</p>
<p dir="rtl" style="text-align:right">כעת נגרום לתוכניות לבחור בין מקרים פיסיקליים נבדלים. לפני שאפשר לסמוך על חישוב, עלינו לדעת איזה סוג של ערך אנו מחזיקים, להמיר במכוון, ולהצהיר על התחום שבו מודל תקף.</p>
<h2 id="לערכים-יש-טיפוסים">לערכים יש טיפוסים</h2>
<p dir="rtl" style="text-align:right">ערכים ב-<code dir="ltr">Julia</code> נושאים טיפוס. מספרים שלמים, מספרי נקודה צפה, טקסט וערכי אמת משמשים למטרות שונות.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [23]:
println(3, "  ->  ", typeof(3))
println(3.0, "  ->  ", typeof(3.0))
println("3", "  ->  ", typeof("3"))
println(3 < 4, "  ->  ", typeof(3 < 4))

3  ->  

Int64
3.0  ->  Float64
3  ->  String
true  ->  Bool


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">3  -&gt;  Int64
3.0  -&gt;  Float64
3  -&gt;  String
true  -&gt;  Bool</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">הטיפוס השלם המדויק יכול להיות תלוי במחשב; ההבחנה החשובה היא בין <code dir="ltr">Int</code>, נקודה צפה, <code dir="ltr">String</code> ו-<code dir="ltr">Bool</code>.</p>
<p dir="rtl" style="text-align:right">מספר שלם ומספר נקודה צפה יכולים להתערבב באריתמטיקה רגילה; <code dir="ltr">Julia</code> מקדמת את התוצאה לפי הצורך. טקסט אינו הופך בשקט למספר. אנו בוחרים המרות במפורש בשני הכיוונים.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [24]:
println(2 + 0.5)
text = "12.5"
mass = parse(Float64, text)   # kg
println(mass + 0.5)
println(string(mass, " kg"))
println(Int(7.0))

2.5
13.0
12.5 kg
7


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">2.5
13.0
12.5 kg
7</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">parse(Float64, "twelve")</code> מעלה שגיאה משום שהטקסט אינו ספרה של נקודה צפה. גם <code dir="ltr">Int(7.2)</code> מעלה שגיאה: המרה ל-<code dir="ltr">Int</code> מותרת רק כאשר הערך שלם בדיוק. זה מגן עלינו מפני השלכת מידע בשקט.</p>
<h2 id="בוליאנים-והחלטות-הדרגתיות">בוליאנים והחלטות הדרגתיות</h2>
<p dir="rtl" style="text-align:right">השוואות מפיקות ערכי <code dir="ltr">Bool</code>. נשאל תחילה שאלה אחת, אחר כך נבחר בין שני מקרים, ולבסוף נוסיף מקרה שלישי עם <code dir="ltr">elseif</code>.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [25]:
temperature = -2.0   # C
println(temperature < 0)
if temperature < 0
    println("below freezing")
else
    println("not below freezing")
end

true


below freezing


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">true
below freezing</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [26]:
if temperature < 0
    state = "ice"
elseif temperature == 0
    state = "melting-point"
else
    state = "liquid"
end
println(state)

ice


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">ice</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">ice</code>. בדיוק ענף אחד משרשרת <code dir="ltr">if</code>/<code dir="ltr">elseif</code>/<code dir="ltr">else</code> רץ. <code dir="ltr">=</code> מציב; <code dir="ltr">==</code> משווה.</p>
<p dir="rtl" style="text-align:right">אנו משלבים תנאים עם <code dir="ltr">&amp;&amp;</code> (שניהם), <code dir="ltr">||</code> (לפחות אחד) ו-<code dir="ltr">!</code> (לא). בדיקת תקינות שואלת לעתים קרובות אם ערך הוא מספר סופי. <code dir="ltr">isfinite(value)</code> מחזירה <code dir="ltr">true</code> עבור מספרים סופיים רגילים ו-<code dir="ltr">false</code> עבור <code dir="ltr">NaN</code>, <code dir="ltr">Inf</code> או <code dir="ltr">-Inf</code>, שאינם מדידות שמישות. אפשר לשלב בדיקה זו עם תנאי פיסיקלי:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [27]:
is_finite_positive = isfinite(2.0) && 2.0 > 0
println(is_finite_positive)
println(!is_finite_positive || 2.0 == 2)

true
true


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">true
true</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="זוויות-לוגיקה-ואימות-לפני-חישוב">זוויות, לוגיקה ואימות לפני חישוב</h2>
<h4 id="תנאי-הוא-שאלה-מתמטית-מדויקת">תנאי הוא שאלה מתמטית מדויקת</h4>
<p dir="rtl" style="text-align:right"><code dir="ltr">a &amp;&amp; b</code> אמיתי רק כאשר שני התנאים אמיתיים; <code dir="ltr">a || b</code> אמיתי כאשר לפחות אחד אמיתי. <code dir="ltr">xor(a,b)</code> אמיתי כאשר בדיוק אחד אמיתי. הכללים מסייעים לבחירת מקרים זרים, אך המודל חייב לקבוע בנפרד מה קורה בשוויון. אי-שוויון משורשר כמו <span dir="ltr" class="math inline">\(0\le\theta&lt;90^\circ\)</span> דורש ששני הגבולות יתקיימו; אפשר לכתוב שתי השוואות המחוברות ב-<code dir="ltr">&amp;&amp;</code>.</p>
<p dir="rtl" style="text-align:right">לזווית דרושה יחידה מפורשת. <code dir="ltr">sin</code> ו-<code dir="ltr">cos</code> מצפות לרדיאנים, ואילו <code dir="ltr">sind</code> ו-<code dir="ltr">cosd</code> מצפות למעלות; <code dir="ltr">rad2deg</code> ממירה זווית מרדיאנים למעלות. הצורה הדו-ארגומנטית <code dir="ltr">atan(y, x)</code> מחזירה את הזווית של הנקודה <span dir="ltr" class="math inline">\((x,y)\)</span> תוך התחשבות ברביע: היא משתמשת בשני הסימנים, ולא רק ביחס ביניהם, ולכן יכולה להבחין בין כיוונים מנוגדים. דוגמה קטנה לכל אחת מהן היא:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [28]:
println(true && false)
println(true || false)
println(xor(true, true))
println(xor(true, false))
println(sin(pi / 6))
println(sind(30.0))
println(round(rad2deg(atan(1.0, -1.0)); digits=1))

false
true
false
true
0.49999999999999994
0.5
135.0


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">false
true
false
true
0.49999999999999994
0.5
135.0</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">sin</code> מקבלת רדיאנים ו-<code dir="ltr">sind</code> מקבלת מעלות. <code dir="ltr">atan(y,x)</code> משתמשת בשני הסימנים כדי לבחור רביע נכון: הנקודה <span dir="ltr" class="math inline">\((-1,1)\)</span> נמצאת בזווית <span dir="ltr" class="math inline">\(135^\circ\)</span>, לא <span dir="ltr" class="math inline">\(-45^\circ\)</span>. בראשית אין כיוון פיסיקלי מוגדר, ולכן יישום חייב לקבוע מוסכמה. הבדלים זעירים בין חישובי נקודה צפה שקולים אינם הבדל בפיסיקה.</p>
<h4 id="המרת-טקסט-ואימות-פיסיקלי-הם-שלבים-נפרדים">המרת טקסט ואימות פיסיקלי הם שלבים נפרדים</h4>
<p dir="rtl" style="text-align:right"><code dir="ltr">readline()</code> מחזירה טקסט. <code dir="ltr">parse(Float64, text)</code> הופכת טקסט מספרי תקין למספר; טקסט פגום גורם לשגיאה. המרה מוצלחת אינה מספיקה: <code dir="ltr">NaN</code> ו-<code dir="ltr">Inf</code> אינם מדידות סופיות, וגם מסה שלילית סופית נמצאת מחוץ למודל. נאמת לפני שורש, חלוקה או בנייה טריגונומטרית. <code dir="ltr">throw(ArgumentError("message"))</code> עוצרת את החישוב; הדפסת הודעה בלבד אינה עוצרת אותו.</p>
<p dir="rtl" style="text-align:right">לפני בדיקת תחום פיסיקלי, אנו בודקים שהערך סופי: <code dir="ltr">isfinite(value)</code> הוא <code dir="ltr">true</code> רק עבור מספרים סופיים רגילים ו-<code dir="ltr">false</code> עבור <code dir="ltr">NaN</code> או <code dir="ltr">Inf</code>. המקטע הבא ממיר טקסט ומאמת את התוצאה כמסה:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [29]:
text = "2.5"  # reproducible example of text supplied by readline()
mass = parse(Float64, text)   # kg
if !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite and positive"))
end
println("mass = ", mass, " kg")

mass = 2.5 kg


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">mass = 2.5 kg</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">הדוגמאות הקודמות של <code dir="ltr">tryparse</code> ו-<code dir="ltr">nothing</code> נשמרות לקריאת רשות על טיפול בקלט. מסלול החובה השבוע משתמש ב-<code dir="ltr">parse</code> ובבדיקות תחום מפורשות; אין צורך בהגדרת פונקציה באף אחד מהשלבים.</p>
<h2 id="מודל-חיכוך-בחלקים-על-מישור-משופע">מודל חיכוך בחלקים על מישור משופע</h2>
<h4 id="כוחות-סימנים-והנחות">כוחות, סימנים והנחות</h4>
<p dir="rtl" style="text-align:right">נבחר <span dir="ltr" class="math inline">\(+x\)</span> במעלה המישור הקבוע ו-<span dir="ltr" class="math inline">\(+y\)</span> ניצב לו. עבור גוף בעל מסה <span dir="ltr" class="math inline">\(m\)</span> (ק"ג), זווית <span dir="ltr" class="math inline">\(\theta\)</span> (רדיאנים ב-<code dir="ltr">sin</code>/<code dir="ltr">cos</code>) וגודל כבידה חיובי <span dir="ltr" class="math inline">\(g\)</span> (m/s<span dir="ltr" class="math inline">\(^2\)</span>), לרכיב הכבידה במורד המישור יש גודל <span dir="ltr" class="math inline">\(mg\sin\theta\)</span> והכוח הנורמלי הוא <span dir="ltr" class="math inline">\(N=mg\cos\theta\)</span>. החיכוך הסטטי יכול להתנגד לתנועה הממשמשת במורד עד <span dir="ltr" class="math inline">\(f_{s,\max}=\mu_sN\)</span>. אם אינו מצליח להחזיק, החיכוך הקינטי <span dir="ltr" class="math inline">\(f_k=\mu_kN\)</span> פועל במעלה המישור ו-<span dir="ltr" class="math inline">\(a=(-mg\sin\theta+\mu_kN)/m\)</span>. אנו מניחים חיכוך קולון, מגע קשיח, ללא גרר, ו-<span dir="ltr" class="math inline">\(0\le\theta&lt;90^\circ\)</span>. המודל אינו מתאר גלגול, מגע משתנה או חיכוך תלוי מהירות. ב-<span dir="ltr" class="math inline">\(\theta=0\)</span> החיכוך הסטטי הנדרש הוא אפס; אם <span dir="ltr" class="math inline">\(\mu_k=0\)</span>, ההחלקה נותנת <span dir="ltr" class="math inline">\(a=-g\sin\theta\)</span>.</p>
<p dir="rtl" style="text-align:right">אנו מאמתים לפני שימוש בקלט. <code dir="ltr">throw(ArgumentError(...))</code> עוצר מיד את החישוב ומודיע לקורא שהתנאי המתועד לקלט לא התקיים. זה שונה מ-<code dir="ltr">println</code>, שמדפיס טקסט בלבד וממשיך. בדיקת השייכות <code dir="ltr">value isa T</code> שואלת אם ערך שייך לטיפוס <code dir="ltr">T</code>; <code dir="ltr">Real</code> הוא שם המשפחה של טיפוסי המספרים הממשיים ב-<code dir="ltr">Julia</code>, הכוללת מספרים שלמים וערכי נקודה צפה רגילים. <code dir="ltr">Bool</code> דורש זהירות מיוחדת כאן: משום ש-<code dir="ltr">true isa Real</code> גם הוא <code dir="ltr">true</code>, בדיקה ל-<code dir="ltr">Real</code> בלבד תקבל <code dir="ltr">true</code> כמסה, ולכן נדחה אותו במפורש.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>לפני ההרצה:</strong> נחזה את הערך, הטקסט או הגרף המתוארים לעיל. זו הדוגמה המתאימה מהפרק הכתוב.</p>
</div>

In [30]:
mass = 2.0      # kg
angle = 30.0    # deg
 μ_s = 0.60
 μ_k = 0.40
g = 9.81        # m/s^2
if !(mass isa Real) || mass isa Bool || !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite, positive, and non-Boolean"))
elseif !isfinite(angle) || !(0 <= angle < 90) || μ_k < 0 || μ_k > μ_s
    throw(ArgumentError("use 0 <= angle < 90 and 0 <= μ_k <= μ_s"))
end
 θ = deg2rad(angle)   # rad
normal = mass * g * cos(θ)   # N
needed_static = mass * g * sin(θ)   # N
if needed_static <= μ_s * normal
    state = "stuck"
    acceleration = 0.0   # m/s^2
else
    state = "sliding"
    acceleration = (-needed_static + μ_k * normal) / mass   # m/s^2
end
println(state, ", a = ", acceleration, " m/s^2")

stuck, a = 0.0 m/s^2


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>בדיקה:</strong> נשווה את התוצאה להסבר שבהמשך. תא מציג אוטומטית רק את הביטוי האחרון; <code dir="ltr">println</code> מדפיסה טקסט במפורש. לאחר שינוי קלט נריץ שוב את התאים התלויים בו.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">פלט הייחוס בפרק הכתוב, להשוואה לפלט שהתקבל בפועל:</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">stuck, a = 0.0 m/s^2</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">stuck, a = 0.0 m/s^2</code>. כאן <span dir="ltr" class="math inline">\(mg\sin30^\circ=9.81\)</span> N ו-<span dir="ltr" class="math inline">\(\mu_sN\approx10.19\)</span> N, ולכן החיכוך הסטטי מספיק. אם נשנה רק את <code dir="ltr">μ_s</code> ל-<code dir="ltr">0.40</code>, התוצאה תהיה <code dir="ltr">sliding</code> עם תאוצה שלילית: במורד המישור לפי המוסכמה שהצהרנו. שוויון שייך לענף ה-<code dir="ltr">stuck</code>, מוסכמת הסף המפורשת שלנו.</p>
<h2 id="תרגול-משיכה-יכולה-להפוך-את-כיוון-התנועה-הצפוי">תרגול: משיכה יכולה להפוך את כיוון התנועה הצפוי</h2>
<p dir="rtl" style="text-align:right">נוסיף למודל המישור כוח משיכה ידוע <span dir="ltr" class="math inline">\(P\)</span> כלפי מעלה. לפני בחירת סימן החיכוך נחשב את הנטייה ללא חיכוך, <span dir="ltr" class="math inline">\(T=P-mg\sin\theta\)</span>. כאשר <span dir="ltr" class="math inline">\(T\)</span> חיובי, התנועה הצפויה היא במעלה המישור; כאשר הוא שלילי, התנועה הצפויה היא במורד. החיכוך מתנגד לנטייה זו, וחיכוך סטטי יכול להחזיק את הגוף כאשר <span dir="ltr" class="math inline">\(|T|\le\mu_sN\)</span>.</p>
<p dir="rtl" style="text-align:right">נגיע למפגש כשאנו יודעים לקרוא שרשרת תנאים ולהמיר מעלות לרדיאנים. בחלק המודרך הקצר נחזה כיוון, נעקוב אחר ענף אחד ונבדוק את <span dir="ltr" class="math inline">\(T=0\)</span>. לאחר מכן נתחיל את מטלת מעבר הקיר. נתרגל שיטה חוזרת: נאמת קלט, ננסח תנאי פיסיקלי, נחשב את המקרה המתאים ונבדוק את גבולו.</p>
<h2 id="מטלת-החובה-האם-קליע-עובר-מעל-קיר">מטלת החובה: האם קליע עובר מעל קיר?</h2>
<p dir="rtl" style="text-align:right">נשגר מגובה הקרקע במהירות <span dir="ltr" class="math inline">\(v_0\)</span> ובזווית <span dir="ltr" class="math inline">\(\theta\)</span>, עם גודל כבידה <span dir="ltr" class="math inline">\(g&gt;0\)</span>. נגדיר <span dir="ltr" class="math inline">\(+x\)</span> לכיוון הקיר ו-<span dir="ltr" class="math inline">\(+y\)</span> מעלה. במודל של כבידה אחידה ללא גרר, <span dir="ltr" class="math display">\[v_x=v_0\cos\theta,\quad t_w=x_w/v_x,\quad y_w=v_0\sin\theta\,t_w-\tfrac12gt_w^2,\quad v_{y,w}=v_0\sin\theta-gt_w.\]</span> נדרוש <span dir="ltr" class="math inline">\(v_x&gt;0\)</span> לפני החלוקה. נשווה את <span dir="ltr" class="math inline">\(y_w\)</span> לגובה הקיר ונשתמש ב-<span dir="ltr" class="math inline">\(v_{y,w}\)</span> לסיווג עלייה או ירידה. שוויון בגובה נחשב מעבר; מהירות אנכית אפס שייכת לקטגוריית הירידה לפי המוסכמה שפורסמה. גובה התנועה החופשית המחושב עשוי להיות שלילי: עבור קיר שגובהו אינו שלילי, פירושו אי-מעבר ולא המשך פיסיקלי דרך הקרקע.</p>
<p dir="rtl" style="text-align:right">נכתוב תחזית למקרה רגיל ולמקרה גבול, נממש את הקטע הסקלרי ונאמת את התחום המותר. נשווה מחרוזות וערכים בוליאניים בדיוק, וגבהים בקירוב לפי הסבולת שפורסמה. נסביר את ההבדל בין החלטת המעבר לסיווג התנועה: אפשר לעבור את הקיר גם בירידה. דף המטלה הוא המקום שבו מפורטים שמות הקלט, הבדיקות, מוסכמת הפונקציה העוטפת וההגשה; אין צורך בלולאות או בהגדרת פונקציות.</p>
<h2 id="בונוס-לבחירה-יחידות-ופאזורים-בזרם-חילופין">בונוס לבחירה: יחידות ופאזורים בזרם חילופין</h2>
<p dir="rtl" style="text-align:right">הבונוס הוא רשות, אינו תנאי להמשך ואינו נבחן. הוא מזכה ב<strong>שליש נקודה בציון הסופי</strong>.</p>
<p dir="rtl" style="text-align:right">הקריאה העצמאית מפתחת מספרים מרוכבים, גודל, פאזה, צמוד וקשר אוילר לפני השימוש ב-<code dir="ltr">cis</code>. נבחר משרעות שיא ואת מוסכמת הזמן <span dir="ltr" class="math inline">\(\exp(+i\omega t)\)</span>, נקבע את פאזת המתח לאפס, נבנה <span dir="ltr" class="math inline">\(Z=|Z|\exp(i\phi_Z)\)</span> ונחשב <span dir="ltr" class="math inline">\(I=V/Z\)</span>. הזרם המדיד הוא החלק הממשי של <span dir="ltr" class="math inline">\(I\exp(i\omega t)\)</span>; המספר המרוכב הוא ייצוג ולא זרם נמדד נוסף.</p>
<p dir="rtl" style="text-align:right">נעקוב אחר דוגמאות <code dir="ltr">Unitful</code>, נוודא שלמתח חלקי התנגדות יש יחידות זרם ונבדוק נגד בעל פאזה אפס לפני מקרה של רבע מחזור. נגיש קטע המציב זרם מרוכב באמפר ונריץ את בדיקות היחידות השקולות והקלט הפסול. נסביר מדוע פאזה חיובית של העכבה נותנת פאזה שלילית של הזרם. המודל לינארי, בעל תדירות אחת ובמצב מתמיד; מעברי מיתוג ורכיבים לא-לינאריים אינם מתוארים בו.</p>
<h2 id="להעמקה-החזרה-פנימית-מלאה">להעמקה: החזרה פנימית מלאה</h2>
<p dir="rtl" style="text-align:right">נשתמש במתכון המסופק לחוק סנל <span dir="ltr" class="math inline">\(n_1\sin\theta_1=n_2\sin\theta_2\)</span>, עם מקדמי שבירה חיוביים וזוויות הנמדדות מהנורמל לממשק. עבור <span dir="ltr" class="math inline">\(n_1&gt;n_2\)</span>, זווית פגיעה הגדולה מהזווית הקריטית <span dir="ltr" class="math inline">\(\theta_c=\arcsin(n_2/n_1)\)</span> גורמת להחזרה פנימית מלאה. נסווג דוגמאות מתחת לסף, על הסף ומעליו לפני חישוב סינוס הפוך; הפעלת <code dir="ltr">asin</code> על מספר גדול מאחת ללא בדיקה אינה חישוב פיסיקלי תקין.</p>
<p dir="rtl" style="text-align:right">נשרטט החזרות אידיאליות חוזרות בסיב ישר ונשנה את זווית הכניסה. נרשום אילו בחירות נשארות מונחות בסיב ונסביר מהו הנורמל בכל ממשק. לולאת עקיבה מסופקת, אם נשתמש בה, היא הצצה לבחירה ולא חומר חובה בשבוע 2. אופני גל, נפיצה, הפסדי כיפוף ובליעה דורשים פיסיקה נוספת. החקירה נפרדת מבונוס הפאזורים, אינה נבחנת ואינה תנאי להמשך.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="סיכום-ופרשנות">סיכום ופרשנות</h2>
<p dir="rtl" style="text-align:right">נסביר תוצאה אחת במילים, נציין יחידות או הנחות כשצריך ונזהה בדיקה שעשויה לחשוף טעות. תיאורי התרגול, מטלת החובה, הבונוס וההעמקה שלעיל מכוונים להמשך. רקע לבחירה אינו הופך לתנאי למבחן.</p>
</div>